pkg/deploy/api/validation/validation.go

```diff
@@ -39,7 +39,18 @@ func ValidateDeploymentConfigSpec(spec deployapi.DeploymentConfigSpec) field.Err
 	if spec.Template == nil {
 		allErrs = append(allErrs, field.Required(specPath.Child("template"), ""))
 	} else {
+		// rather than handleEmptyImageReferences returning an array of changed containers, PR review deemed it more
+		// efficient to build a list of the original container to image mappings, and then just restore all of them (and hence,
+		// even the ones not modified)
+		originalContainerImages := make(map[string]string)
+		for _, container := range spec.Template.Spec.Containers {
+			originalContainerImages[container.Name] = container.Image
+		}
+		handleEmptyImageReferences(spec.Template, spec.Triggers)
```

Comment:
a pair of functions that replaces and returns the original container image names would feel safer to me:
```
func getContainerImageNames(template *kapi.PodTemplateSpec) []string {
  names := make([]string, len(template.Containers))
  for i := range template.Containers {
    names[i] = template.Containers[i].Image
  }
  return names
}

func setContainerImageNames(template *kapi.PodTemplateSpec, names []string) {
  for i := range template.Containers {
    template.Containers[i].Image = names[i]
  }
}
```

then you could use it like:
```
originalNames := getContainerImageNames(spec.Template)
defer setContainerImageNames(spec.Template, originalNames)
```

and continue on, overriding image names however you like